File: fons/norma-py/toml.py
"""toml.py - TOML Encoding/Decoding Implementation

Native Python implementation of the TOML interface.
Uses tomllib (read) and tomli_w (write) for Python 3.11+.

Note: TOML root must be a table (object), not array or primitive.

Verb meanings:
  - pange (compose): serialize table to TOML string
  - solve (untangle): parse TOML string to table
  - tempta (try): attempt to parse, return None on error
"""

import tomllib
from typing import Any, Optional

try:
    import tomli_w
    _HAS_WRITER = True
except ImportError:
    _HAS_WRITER = False


# =============================================================================
# SERIALIZATION
# =============================================================================

def pange(valor: Any) -> str:
    """Serialize table to TOML string."""
    if not _HAS_WRITER:
        raise RuntimeError("tomli_w package required for TOML serialization")
    return tomli_w.dumps(valor)


# =============================================================================
# PARSING
# =============================================================================

def solve(toml_str: str) -> Any:
    """Parse TOML string to table (raises on error)."""
    return tomllib.loads(toml_str)


def tempta(toml_str: str) -> Optional[Any]:
    """Attempt to parse TOML string (returns None on error)."""
    try:
        return tomllib.loads(toml_str)
    except (ValueError, tomllib.TOMLDecodeError):
        return None


# =============================================================================
# TYPE CHECKING
# =============================================================================

def est_nihil(valor: Any) -> bool:
    """Check if value is null (TOML doesn't have null, always False)."""
    return False


def est_bivalens(valor: Any) -> bool:
    """Check if value is boolean."""
    return isinstance(valor, bool)


def est_textus(valor: Any) -> bool:
    """Check if value is string."""
    return isinstance(valor, str)


def est_integer(valor: Any) -> bool:
    """Check if value is integer."""
    return isinstance(valor, int) and not isinstance(valor, bool)


def est_fractus(valor: Any) -> bool:
    """Check if value is float."""
    return isinstance(valor, float)


def est_tempus(valor: Any) -> bool:
    """Check if value is datetime."""
    from datetime import datetime, date, time
    return isinstance(valor, (datetime, date, time))


def est_lista(valor: Any) -> bool:
    """Check if value is array."""
    return isinstance(valor, list)


def est_tabula(valor: Any) -> bool:
    """Check if value is table."""
    return isinstance(valor, dict)


# =============================================================================
# VALUE EXTRACTION
# =============================================================================

def ut_textus(valor: Any, def_val: str) -> str:
    """Extract as string with default."""
    return valor if isinstance(valor, str) else def_val


def ut_numerus(valor: Any, def_val: float) -> float:
    """Extract as number with default."""
    if isinstance(valor, float):
        return valor
    if isinstance(valor, int) and not isinstance(valor, bool):
        return float(valor)
    return def_val


def ut_bivalens(valor: Any, def_val: bool) -> bool:
    """Extract as boolean with default."""
    return valor if isinstance(valor, bool) else def_val


# =============================================================================
# VALUE ACCESS
# =============================================================================

def cape(valor: Any, clavis: str) -> Optional[Any]:
    """Get value by key (returns None if missing)."""
    if isinstance(valor, dict):
        return valor.get(clavis)
    return None


def carpe(valor: Any, index: int) -> Optional[Any]:
    """Pluck value by array index (returns None if out of bounds)."""
    if isinstance(valor, list) and 0 <= index < len(valor):
        return valor[index]
    return None


def inveni(valor: Any, via: str) -> Optional[Any]:
    """Find value by dotted path (returns None if not found)."""
    parts = via.split(".")
    current = valor

    for part in parts:
        if current is None:
            return None
        if isinstance(current, dict):
            current = current.get(part)
        elif isinstance(current, list):
            try:
                idx = int(part)
                if 0 <= idx < len(current):
                    current = current[idx]
                else:
                    return None
            except ValueError:
                return None
        else:
            return None

    return current

File: runtimes/norma-rs/toml.rs
//! toml.rs - TOML Encoding/Decoding Implementation
//!
//! Native Rust implementation of the TOML interface.
//! Uses the toml crate.
//!
//! Note: TOML root must be a table (object), not array or primitive.
//!
//! Verb meanings:
//!   - pange (compose): serialize table to TOML string
//!   - solve (untangle): parse TOML string to table
//!   - tempta (try): attempt to parse, return None on error

use toml::Value;

// =============================================================================
// SERIALIZATION
// =============================================================================

/// Serialize table to TOML string
pub fn pange(valor: &Value) -> String {
    toml::to_string(valor).expect("failed to serialize TOML")
}

// =============================================================================
// PARSING
// =============================================================================

/// Parse TOML string to table (panics on error)
pub fn solve(toml_str: &str) -> Value {
    toml::from_str(toml_str).expect("failed to parse TOML")
}

/// Attempt to parse TOML string (returns None on error)
pub fn tempta(toml_str: &str) -> Option<Value> {
    toml::from_str(toml_str).ok()
}

// =============================================================================
// TYPE CHECKING
// =============================================================================

/// Check if value is null (TOML doesn't have null, always false)
pub fn est_nihil(_valor: &Value) -> bool {
    false // TOML has no null type
}

/// Check if value is boolean
pub fn est_bivalens(valor: &Value) -> bool {
    valor.is_bool()
}

/// Check if value is string
pub fn est_textus(valor: &Value) -> bool {
    valor.is_str()
}

/// Check if value is integer
pub fn est_integer(valor: &Value) -> bool {
    valor.is_integer()
}

/// Check if value is float
pub fn est_fractus(valor: &Value) -> bool {
    valor.is_float()
}

/// Check if value is datetime
pub fn est_tempus(valor: &Value) -> bool {
    valor.is_datetime()
}

/// Check if value is array
pub fn est_lista(valor: &Value) -> bool {
    valor.is_array()
}

/// Check if value is table
pub fn est_tabula(valor: &Value) -> bool {
    valor.is_table()
}

// =============================================================================
// VALUE EXTRACTION
// =============================================================================

/// Extract as string with default
pub fn ut_textus(valor: &Value, def_val: &str) -> String {
    valor
        .as_str()
        .map(|s| s.to_string())
        .unwrap_or_else(|| def_val.to_string())
}

/// Extract as number with default
pub fn ut_numerus(valor: &Value, def_val: f64) -> f64 {
    valor
        .as_float()
        .or_else(|| valor.as_integer().map(|i| i as f64))
        .unwrap_or(def_val)
}

/// Extract as boolean with default
pub fn ut_bivalens(valor: &Value, def_val: bool) -> bool {
    valor.as_bool().unwrap_or(def_val)
}

// =============================================================================
// VALUE ACCESS
// =============================================================================

/// Get value by key (returns integer 0 as sentinel for missing - TOML has no null)
pub fn cape(valor: &Value, clavis: &str) -> Option<Value> {
    valor.get(clavis).cloned()
}

/// Pluck value by array index (returns None if out of bounds)
pub fn carpe(valor: &Value, index: usize) -> Option<Value> {
    valor.get(index).cloned()
}

/// Find value by dotted path (returns None if not found)
pub fn inveni(valor: &Value, via: &str) -> Option<Value> {
    let parts: Vec<&str> = via.split('.').collect();
    let mut current = valor;

    for part in parts {
        match current {
            Value::Table(map) => {
                current = match map.get(part) {
                    Some(v) => v,
                    None => return None,
                };
            }
            Value::Array(arr) => {
                let idx: usize = match part.parse() {
                    Ok(i) => i,
                    Err(_) => return None,
                };
                current = match arr.get(idx) {
                    Some(v) => v,
                    None => return None,
                };
            }
            _ => return None,
        }
    }

    Some(current.clone())
}
